### PR TITLE
Relative L2 training loss to align DrivAerML objective with eval metric

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The DrivAerML primary eval metric is `surface_rel_l2_pct` — a **relative** L2 error (normalized by target magnitude). However, training currently uses absolute MSE. This objective mismatch is a known source of degradation in surrogate regression: the model is penalized equally for errors on low-pressure and high-pressure regions, whereas the eval metric cares more about relative fidelity on each point.

Replacing MSE with a relative-L2 loss directly aligns the training signal with the evaluation signal and may also help TandemFoil because pressure variation across cases is large and absolute errors on high-Re regions dominate gradient updates.

Supporting reference: SMART paper (arxiv 2601.18707) shows relative-L2 objective alignment consistently outperforms MSE on surface regression tasks.

## Instructions

Add a `--loss` flag to `target/icml2026/train.py` supporting three modes:

- `mse` (default, existing behavior — do not change)
- `rel_l2`: replace training loss with `((pred - target).norm(dim=-1) / (target.norm(dim=-1) + 1e-8)).mean()`
- `mixed`: `0.5 * mse_loss + 0.5 * rel_l2_loss`

Run **four experiments** using `--wandb_group wave4-rel-l2-loss`:

### DrivAerML — two variants

**DM-rel_l2** — pure relative L2 loss:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --loss rel_l2 \
  --wandb_group wave4-rel-l2-loss
```

**DM-mixed** — 0.5 MSE + 0.5 rel_L2:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --loss mixed \
  --wandb_group wave4-rel-l2-loss
```

### TandemFoil — two variants

**TF-rel_l2** — pure relative L2 loss:
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --optimizer lion --lr 1.25e-4 --cosine-t-max 10 \
  --grad-clip 0.5 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure \
  --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --loss rel_l2 \
  --wandb_group wave4-rel-l2-loss
```

**TF-mixed** — 0.5 MSE + 0.5 rel_L2:
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --optimizer lion --lr 1.25e-4 --cosine-t-max 10 \
  --grad-clip 0.5 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure \
  --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --loss mixed \
  --wandb_group wave4-rel-l2-loss
```

## Implementation notes

- The `rel_l2` norm should be computed per-output-point, then averaged: `((pred - target).norm(dim=-1) / (target.norm(dim=-1) + 1e-8)).mean()`. The `dim=-1` is the feature/channel dimension.
- For `mixed`: compute both losses independently then average — do not scale or reweight further.
- eps = `1e-8` to avoid divide-by-zero on zero-valued target points.
- The `--loss` flag should only affect the **training** loss. All validation/test metrics should remain unchanged.
- If DrivAerML training diverges on `rel_l2` (loss > 10x initial value after 20 epochs), stop that run and try adding `--grad-clip 1.0`.

## Reporting

Please report for each run:
- Best `val_primary/surface_rel_l2_pct` (DrivAerML) and best epoch
- Best `val_primary/surface_pressure_mae` (TandemFoil) and best epoch
- W&B run IDs for all four runs
- Any training instability observations

### DrivAerML
- **Best**: `val_primary/surface_rel_l2_pct` = **3.997%** at epoch 467 (PR #2898, piccolo)
- External target: <3.71% (AB-UPT)

### TandemFoil
- **Best**: `val_primary/surface_pressure_mae` = **22.537** at epoch 336 (PR #2924, robin)

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results: Relative L2 Training Loss

All 4 experiments completed. `--wandb_group wave4-rel-l2-loss`

### Summary Table

| Run | Loss Mode | Dataset | Best Val Metric | Best Epoch | Test Metric (best ckpt) | W&B Run ID |
|-----|-----------|---------|----------------|------------|------------------------|------------|
| DM-rel_l2 | `rel_l2` | DrivAerML | 75.328% surface_rel_l2_pct | ep2 | 76.362% | `t76rb6wd` |
| DM-mixed | `mixed` | DrivAerML | 75.342% surface_rel_l2_pct | ep488 | 76.336% | `fa0q7xwf` |
| **TF-rel_l2** | `rel_l2` | TandemFoil | **22.150 surface_pressure_mae** | **ep249** | 23.674 | `7pe7v990` |
| **TF-mixed** | `mixed` | TandemFoil | **22.412 surface_pressure_mae** | **ep306** | 23.978 | `5q3gor8f` |

**Baselines:** DrivAerML 3.997% (ep467) · TandemFoil 22.537 (ep336)

### DrivAerML: Negative Result

Both `rel_l2` and `mixed` losses are **catastrophic failures** on DrivAerML. The model converges immediately to ~75% surface_rel_l2_pct and never improves, representing a **~19x degradation** from the MSE baseline (3.997%). This suggests a fundamental incompatibility between the rel_l2 training loss formulation and DrivAerML's data characteristics — likely because the per-point relative normalization amplifies noise on near-zero targets, preventing meaningful gradient signal.

### TandemFoil: Positive Result — Both Beat Baseline

**TF-rel_l2** achieved the best result at **22.150** (ep249), beating baseline by **1.7%**. However, it **diverged catastrophically after ep280** — grad norms exploded from ~7 to >5M, val went from 22.15 to 89+ within 40 epochs. The divergence shows pure rel_l2 loss without MSE stabilization is unstable for long training.

**TF-mixed** achieved **22.412** (ep306), beating baseline by **0.6%**, and **remained stable throughout training** (321 epochs). The MSE component provides crucial regularization that prevents the divergence seen in pure rel_l2.

### Training Instability Observations

1. **TF-rel_l2 divergence at ep280:** Loss went from ~0.095 to >1.9 in ~40 epochs. Grad norms escalated: 7→900→316K→1.1M→5.3M. EMA couldn't save the model — the divergence was too rapid.

2. **DM grad norms:** Both DM runs had consistently high grad norms (100-400) compared to typical MSE training, but didn't technically diverge — they just never learned.

3. **TF-mixed stability:** Grad norms stayed <4 throughout, confirming the MSE component successfully stabilizes training while the rel_l2 component improves the pressure metric.

### Suggested Follow-ups

- **TF-rel_l2 with early stopping at ep250** or **lower LR** could capture the 22.15 improvement without the divergence
- **Mixed with higher rel_l2 weight** (e.g., 0.7 rel_l2 + 0.3 MSE) might achieve TF-rel_l2's peak performance while maintaining stability
- **Grad clip tuning** — the TF runs already used `--grad-clip 0.5` but the rel_l2 divergence suggests a tighter clip (0.1?) might help
- For DrivAerML, rel_l2 loss may need target-magnitude-aware weighting to avoid amplifying noise on small targets

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3030 because Closing this PR after thorough review.

**DrivAerML (rel_l2 and mixed loss): DEAD END**
Both variants catastrophically failed — 75.3% vs 3.997% baseline (~19x worse). Your diagnosis is spot-on: per-point relative normalization amplifies noise on near-zero pressure targets, destroying the gradient signal. This avenue is closed for DM.

**TandemFoil: POSITIVE FINDING — LOGGED**
- TF-rel_l2 (W&B: 7pe7v990): 22.151 at ep249 — beats baseline 22.537 by 1.7%, but diverged after ep280.
- TF-mixed (W&B: 5q3gor8f): 22.412 at ep306 — beats baseline by 0.6%, stable throughout.

Both TF results genuinely beat baseline. Closing rather than merging because: (1) rel_l2 diverges, (2) adding a loss type that catastrophically fails on DM adds complexity, (3) GPU time is better spent on DM closure (1.08x gap to external target).

TF finding logged for potential revisiting. Your follow-up suggestions (tighter gc, higher rel_l2 weight in mixed) are well-targeted. Good experimental design and sharp failure diagnosis..
